libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2598,7 +2644,10 @@ string YulUtilFunctions::updateStorageValueFunction(
 			fromReferenceType->isPointer()
 		).get() == *fromReferenceType, "");
 
-		solAssert(toReferenceType->category() == fromReferenceType->category(), "");
+		if (fromReferenceType->category() == Type::Category::ArraySlice)
```

Comment:
This was interesting, and I am glad you asked.
closestTemporaryType for calldata slices (when converting to storage array) was memory array and there was "hidden" conversion from calldata to memory and then from memory to storage.

Reply:
Interesting indeed! Not sure how that's happening... ``Type::closestTemporaryType`` seems to return ``mobileType``, if `_targetType` is in storage... and ``ArraySliceType::mobileType`` looks to me as if it returns the underlying array type which should be in calldata... so not sure right now what's going on there :-).

Reply:
Indeed, you are right. It was calldata array comming here before, and now, since `closestTemporaryType` is not used anymore here we get actual slice instead. That is why these changes were needed....